File: Components/DevExpressVCL/Demos/ExpressTile Control/Delphi/RealtorWorld/RealtorWorldAgents.pas
unit RealtorWorldAgents;

interface

uses
  Windows, Messages, SysUtils, Variants, Classes, Graphics, Controls, Forms,
  Dialogs, cxGraphics, cxControls, cxLookAndFeels, DB,
  cxLookAndFeelPainters, dxCustomTileControl, dxTileControl, RealtorWorldDM,
  cxStyles, cxCustomData, cxFilter, cxData, cxDataStorage, cxEdit,
  cxDBData, DBClient, cxGridChartView, cxGridDBChartView, cxGridCustomView,
  cxGridCustomTableView, cxGridTableView, cxGridDBTableView, cxClasses,
  cxGridLevel, cxGrid, ExtCtrls, cxTextEdit, cxCurrencyEdit, cxImage,
  cxMemo, cxSplitter, RealtorWorldBaseFrame, MidasLib;

type
  TfrmAgents = class(TfrmBase)
    tcAgents: TdxTileControl;
    pnDetailSite: TPanel;
    cxgDetailChartLevel1: TcxGridLevel;
    cxgDetailChart: TcxGrid;
    cxgDetailInfo: TcxGrid;
    cxGridDBTableView1: TcxGridDBTableView;
    cxGridLevel1: TcxGridLevel;
    cxgDetailChartDBChartView1: TcxGridDBChartView;
    cxGridDBTableView1Address: TcxGridDBColumn;
    cxGridDBTableView1Beds: TcxGridDBColumn;
    cxGridDBTableView1Baths: TcxGridDBColumn;
    cxGridDBTableView1HouseSize: TcxGridDBColumn;
    cxGridDBTableView1Price: TcxGridDBColumn;
    cxGridDBTableView1Photo: TcxGridDBColumn;
    cxgNorthEastSeries: TcxGridDBChartSeries;
    cxgMidWestSeries: TcxGridDBChartSeries;
    cxgSouthSeries: TcxGridDBChartSeries;
    cxgWestSeries: TcxGridDBChartSeries;
    cxSplitter1: TcxSplitter;
    dsChart: TDataSource;
    cdsChart: TClientDataSet;
    cdsChartMidWest: TIntegerField;
    cdsChartNorthEast: TIntegerField;
    cdsChartSouth: TIntegerField;
    cdsChartWest: TIntegerField;
    cdsChartAgentID: TIntegerField;
    cxStyleRepository1: TcxStyleRepository;
    cxStyle1: TcxStyle;
    cdsChartDate: TIntegerField;
    procedure AA(Sender: TObject; var NewWidth, NewHeight: Integer;
      var Resize: Boolean);
    procedure cxSplitter2BeforeClose(Sender: TObject;
      var AllowClose: Boolean);
  private
    procedure InitializeFrame;
    procedure InitializeChartDataSet;
    procedure OnItemClick(Sender: TdxTileControlItem);
  public
    constructor Create(AOwner: TComponent); override;
    procedure SelectItem(APhotoID, AAgentID: Integer); override;
  end;

implementation

{$R *.dfm}

uses
  RealtorWorldMain;

constructor TfrmAgents.Create(AOwner: TComponent);
begin
  inherited Create(AOwner);
  InitializeFrame;
  InitializeChartDataSet;
end;

procedure TfrmAgents.SelectItem(APhotoID, AAgentID: Integer);
var
  I: Integer;
begin
  for I := 0 to tcAgents.Items.Count - 1 do
    if tcAgents.Items[I].Tag = AAgentID then
    begin
      tcAgents.Items[I].MakeVisible;
      tcAgents.Items[I].Click;
      Break;
    end;
end;

procedure TfrmAgents.InitializeFrame;
var
  AItem: TdxTileControlItem;
  dsAgents: TDataSet;
begin
  tcAgents.BeginUpdate;
  try
    dsAgents := DMRealtorWorld.clHomesAndAgents;
    dsAgents.First;
    while not dsAgents.EOF do
    begin
      AItem := tcAgents.CreateItem(True);
      AItem.Glyph.Image.LoadFromFieldValue(dsAgents.FieldByName('Photo').Value);
      AItem.Style := frmRealtorWorld.tlAgents.Style;
      AItem.Glyph.Align := oaMiddleRight;
      AItem.Glyph.Image.Scale(70, 100);
      AItem.Tag := dsAgents.FieldByName('ID').AsInteger;
      AItem.Style.Font.Size := 13;
      AItem.Text1.Value := dsAgents.FieldByName('FirstName').AsString + ' ' + dsAgents.FieldByName('LastName').AsString;
      AItem.Text1.IndentHorz := 10;
      AItem.Text1.IndentVert := 10;
      AItem.Text2.Value := dsAgents.FieldByName('Phone').AsString;
      AItem.Text2.Align := oaTopLeft;
      AItem.Text2.IndentHorz := 10;
      AItem.Text2.IndentVert := 30;
      AItem.OnClick := OnItemClick;
      dsAgents.Next;
    end;
  finally
    tcAgents.EndUpdate;
  end;
end;

procedure TfrmAgents.OnItemClick(Sender: TdxTileControlItem);
begin
  DMRealtorWorld.clHomesDetail.Filter := 'AgentID=' + IntToStr(Sender.Tag);
  cdsChart.Filter := 'AgentID=' + IntToStr(Sender.Tag);
  DMRealtorWorld.clHomesDetail.First;
end;

procedure TfrmAgents.InitializeChartDataSet;
var
  AYear, I: Integer;
begin
  cdsChart.DisableControls;
  try
    for I := 1 to 6 do
      for AYear := 2003 to 2012 do
      begin
       cdsChart.Append;
       cdsChartDate.Value := AYear;
       cdsChartMidWest.Value := Random(20) + 4;
       cdsChartNorthEast.Value := Random(20) + 5;
       cdsChartSouth.Value := Random(20) + 2;
       cdsChartWest.Value := Random(20) + 3;
       cdsChartAgentID.Value := I;
       cdsChart.Post;
      end;
  finally
    cdsChart.EnableControls;
  end;
end;

procedure TfrmAgents.AA(Sender: TObject; var NewWidth, NewHeight: Integer;
  var Resize: Boolean);
begin
    //
end;

procedure TfrmAgents.cxSplitter2BeforeClose(Sender: TObject;
  var AllowClose: Boolean);
begin
  AllowClose := False;
end;

initialization
  RegisterFrame(IDAgents, TfrmAgents);

end.

File: Components/DevExpressVCL/Demos/ExpressTile Control/CBuilder/RealtorWorld/RealtorWorldAgents.cpp
//---------------------------------------------------------------------------

#include <vcl.h>
#pragma hdrstop

#include "RealtorWorldAgents.h"
#include "RealtorWorldMain.h"
#include "RealtorWorldDM.h"
//---------------------------------------------------------------------------
#pragma package(smart_init)
#pragma link "RealtorWorldBaseFrame"
#pragma link "cxControls"
#pragma link "cxClasses"
#pragma link "cxLookAndFeels"
#pragma link "cxGraphics"
#pragma link "cxLookAndFeelPainters"
#pragma link "cxCurrencyEdit"
#pragma link "cxCustomData"
#pragma link "cxData"
#pragma link "cxDataStorage"
#pragma link "cxDBData"
#pragma link "cxEdit"
#pragma link "cxFilter"
#pragma link "cxGraphics"
#pragma link "cxGrid"
#pragma link "cxGridChartView"
#pragma link "cxGridCustomTableView"
#pragma link "cxGridCustomView"
#pragma link "cxGridDBChartView"
#pragma link "cxGridDBTableView"
#pragma link "cxGridLevel"
#pragma link "cxGridTableView"
#pragma link "cxImage"
#pragma link "cxMemo"
#pragma link "cxSplitter"
#pragma link "cxStyles"
#pragma link "dxCustomTileControl"
#pragma link "dxTileControl"
#pragma link "cxNavigator"
#pragma link "dxDateRanges"
#pragma resource "*.dfm"
TfrmAgents *frmAgents;
//---------------------------------------------------------------------------
__fastcall TfrmAgents::TfrmAgents(TComponent* Owner)
	: TfrmBase(Owner)
{
  InitializeFrame();
  InitializeChartDataSet();
}
//---------------------------------------------------------------------------
void __fastcall TfrmAgents::cxSplitter2BeforeClose(TObject *Sender, bool &AllowClose)
{
  AllowClose = false;
}
//---------------------------------------------------------------------------
void __fastcall TfrmAgents::AA(TObject *Sender, int &NewWidth, int &NewHeight, bool &Resize)
{
//
}
//---------------------------------------------------------------------------
void __fastcall TfrmAgents::InitializeFrame()
{
  TdxTileControlItem* AItem;
  TDataSet* dsAgents;
  tcAgents->BeginUpdate();
  __try
  {
    dsAgents = DMRealtorWorld->clHomesAndAgents;
    dsAgents->First();
    while (!dsAgents->Eof)
    {
      AItem = tcAgents->CreateItem(True);
      AItem->Glyph->Image->LoadFromFieldValue(dsAgents->FieldByName("Photo")->Value);
      AItem->Style = frmRealtorWorld->tlAgents->Style;
      AItem->Glyph->Align = oaMiddleRight;
      AItem->Glyph->Image->Scale(70, 100);
      AItem->Tag = dsAgents->FieldByName("ID")->AsInteger;
      AItem->Style->Font->Size = 13;
      AItem->Text1->Value = dsAgents->FieldByName("FirstName")->AsString + " " + dsAgents->FieldByName("LastName")->AsString;
      AItem->Text1->IndentHorz = 10;
      AItem->Text1->IndentVert = 10;
	  AItem->Text2->Value = dsAgents->FieldByName("Phone")->AsString;
      AItem->Text2->Align = oaTopLeft;
      AItem->Text2->IndentHorz = 10;
      AItem->Text2->IndentVert = 30;
	  AItem->OnClick = OnItemClick;
	  dsAgents->Next();
	};
  }
  __finally
  {
    tcAgents->EndUpdate();
  };
}
//---------------------------------------------------------------------------
void __fastcall TfrmAgents::InitializeChartDataSet()
{
  cdsChart->DisableControls();
  __try
  {
	for (int I = 1; I < 7; I++)
	  for (int AYear = 2003; AYear < 2013; AYear++)
	  {
       cdsChart->Append();
       cdsChartDate->Value = AYear;
       cdsChartMidWest->Value = Random(20) + 4;
       cdsChartNorthEast->Value = Random(20) + 5;
       cdsChartSouth->Value = Random(20) + 2;
       cdsChartWest->Value = Random(20) + 3;
       cdsChartAgentID->Value = I;
	   cdsChart->Post();
      };
  }
  __finally
  {
	cdsChart->EnableControls();
  };
}
//---------------------------------------------------------------------------
void __fastcall TfrmAgents::OnItemClick(TdxTileControlItem *Sender)
{
  DMRealtorWorld->clHomesDetail->Filter = "AgentID=" + IntToStr(Sender->Tag);
  cdsChart->Filter = "AgentID=" + IntToStr(Sender->Tag);
  DMRealtorWorld->clHomesDetail->First();
}
//---------------------------------------------------------------------------
void __fastcall TfrmAgents::SelectItem(int APhotoID, int AAgentID)
{
  for (int i = 0; i < tcAgents->Items->Count; i++)
  {
	if (tcAgents->Items->Items[i]->Tag == AAgentID)
	{
	  tcAgents->Items->Items[i]->MakeVisible();
	  tcAgents->Items->Items[i]->Click();
	};
  };
}
//---------------------------------------------------------------------------
